modules/tensor_mechanics/src/postprocessors/AsymptoticExpansionHomogenizationElasticConstants.C

```diff
@@ -0,0 +1,226 @@
+    _grad_disp_y_yz(_subproblem.mesh().dimension() == 3 ? coupledGradient("dy_yz") : _grad_zero),
+    _grad_disp_z_yz(_subproblem.mesh().dimension() == 3 ? coupledGradient("dz_yz") : _grad_zero),
+    _grad_disp_x_zx(_subproblem.mesh().dimension() == 3 ? coupledGradient("dx_zx") : _grad_zero),
+    _grad_disp_y_zx(_subproblem.mesh().dimension() == 3 ? coupledGradient("dy_zx") : _grad_zero),
+    _grad_disp_z_zx(_subproblem.mesh().dimension() == 3 ? coupledGradient("dz_zx") : _grad_zero),
+    _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_base_name + "elasticity_tensor")),
+    _column(getParam<MooseEnum>("column")),
+    _row(getParam<MooseEnum>("row")),
+    _volume(0),
+    _integral_value(0)
+{
+
+  if (_column == 0)
+  {
+    _k = 0;
+    _l = 0;
+  }
+
+  if (_column == 1)
+  {
+    _k = 1;
+    _l = 1;
+  }
+
+  if (_column == 2)
+  {
+    _k = 2;
+    _l = 2;
+  }
+
+  if (_column == 3)
+  {
+    _k = 1;
+    _l = 2;
+  }
+
+  if (_column == 4)
+  {
+    _k = 0;
+    _l = 2;
+  }
+
+  if (_column == 5)
+  {
+    _k = 0;
+    _l = 1;
+  }
+
+  if (_row == 0)
+  {
+    _i = 0;
+    _j = 0;
+  }
+
+  if (_row == 1)
+  {
+    _i = 1;
+    _j = 1;
+  }
+
+  if (_row == 2)
+  {
+    _i = 2;
+    _j = 2;
+  }
+
+  if (_row == 3)
+  {
+    _i = 1;
+    _j = 2;
+  }
+
+  if (_row == 4)
+  {
+    _i = 0;
+    _j = 2;
+  }
+
+  if (_row == 5)
+  {
+    _i = 0;
+    _j = 1;
+  }
+}
+
+void
+AsymptoticExpansionHomogenizationElasticConstants::initialize()
+{
+  _integral_value = 0;
+  _volume = 0;
+}
+
+void
+AsymptoticExpansionHomogenizationElasticConstants::execute()
+{
+  _integral_value += computeIntegral();
+  _volume += _current_elem_volume;
+}
+
+Real
+AsymptoticExpansionHomogenizationElasticConstants::getValue()
+{
+
+  gatherSum(_integral_value);
+  gatherSum(_volume);
+
+  return (_integral_value / _volume);
+}
+
+void
+AsymptoticExpansionHomogenizationElasticConstants::threadJoin(const UserObject & y)
+{
+  const AsymptoticExpansionHomogenizationElasticConstants & pps =
+      dynamic_cast<const AsymptoticExpansionHomogenizationElasticConstants &>(y);
+
+  _integral_value += pps._integral_value;
+  _volume += pps._volume;
+}
+
+Real
+AsymptoticExpansionHomogenizationElasticConstants::computeQpIntegral()
+{
+  const VariableGradient * grad[6][3];
+  grad[0][0] = &_grad_disp_x_xx;
+  grad[0][1] = &_grad_disp_y_xx;
+  grad[0][2] = &_grad_disp_z_xx;
+
+  grad[1][0] = &_grad_disp_x_yy;
+  grad[1][1] = &_grad_disp_y_yy;
+  grad[1][2] = &_grad_disp_z_yy;
+
+  grad[2][0] = &_grad_disp_x_zz;
+  grad[2][1] = &_grad_disp_y_zz;
+  grad[2][2] = &_grad_disp_z_zz;
+
+  grad[3][0] = &_grad_disp_x_yz;
+  grad[3][1] = &_grad_disp_y_yz;
+  grad[3][2] = &_grad_disp_z_yz;
+
+  grad[4][0] = &_grad_disp_x_zx;
+  grad[4][1] = &_grad_disp_y_zx;
+  grad[4][2] = &_grad_disp_z_zx;
+
+  grad[5][0] = &_grad_disp_x_xy;
+  grad[5][1] = &_grad_disp_y_xy;
+  grad[5][2] = &_grad_disp_z_xy;
```

Comment:
I would store the parameters in a 3d array of pointers from the start. Maybe also make it one single coupled variable parameter and use vector coupling.

Reply:
2d! sorry, that was a typo